File: DynamicPrograming/ts/summingSquares.ts
export function summingSquares(n: number): number {
  const memo = new Map<number, number>();

  const dfs = (value: number): number => {
    if (value === 0) return 0;
    if (memo.has(value)) return memo.get(value)!;

    let best = Infinity;
    for (let i = 1; i * i <= value; i += 1) {
      const square = i * i;
      const attempt = 1 + dfs(value - square);
      best = Math.min(best, attempt);
    }

    memo.set(value, best);
    return best;
  };

  return dfs(n);
}

File: DynamicPrograming/go/summing_squares.go
package dynamicprogramming

// SummingSquares returns minimum number of perfect squares that sum to n.
func SummingSquares(n int) int {
	memo := make(map[int]int)

	var dfs func(int) int
	dfs = func(value int) int {
		if value == 0 {
			return 0
		}
		if val, ok := memo[value]; ok {
			return val
		}
		best := maxInt
		for i := 1; i*i <= value; i++ {
			candidate := 1 + dfs(value-i*i)
			if candidate < best {
				best = candidate
			}
		}
		memo[value] = best
		return best
	}

	return dfs(n)
}
